File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* GetKthNode(ListNode* head ,int k){
        k -= 1;
        while(head != NULL && k > 0){
            k--;
            head = head->next;
        }
        return head;
    }
    ListNode* reverse(ListNode* head){
        ListNode* prev = NULL;
        ListNode* curr = head;

        while(curr != NULL){
            ListNode* nextNode = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nextNode;
        }
        return prev;
    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* prevNode = NULL;
        ListNode* temp = head;

        while(temp != NULL){
            ListNode* kthNode = GetKthNode(temp,k);

            if(kthNode == NULL){
                if(prevNode)prevNode->next = temp;
                break;
            }
            ListNode* nextNode = kthNode->next;
            kthNode->next = NULL;
            reverse(temp);
            if(temp == head){
                head = kthNode;
            }
            else{
                prevNode->next = kthNode;
            }
            prevNode = temp;
            temp = nextNode;
        }
        return head;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode GetKthNode(ListNode head, int k){     
        k -= 1;
        while(head != null && k > 0){
            k--;
            head = head.next;
        }
        return head;
    }
    public ListNode reverseLinkList(ListNode head){
        ListNode prev = null;
        ListNode curr = head;

        while(curr != null){
            ListNode nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }
        return prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode prevNode = null;
        ListNode temp = head;

        while(temp != null){
            ListNode KthNode = GetKthNode(temp,k);
            if(KthNode == null){
                if(prevNode != null)prevNode.next = temp;
                break;
            }
            ListNode nextNode = KthNode.next;
            KthNode.next = null;
            reverseLinkList(temp);
            if(temp == head){
                head = KthNode;
            }
            else{
                prevNode.next = KthNode;
            }
            prevNode = temp;
            temp = nextNode;
        }
        return head;
    }
}